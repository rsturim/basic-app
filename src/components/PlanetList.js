import React, { Component } from 'react';

import SwapiService from '../common/SwapiService';

export default class PlanetList extends Component {
    constructor(props) {
        super(props);

        this.state = {
            isLoading: true,
            planets: [],
            lifeCycle: '',
        };
    }

    componentDidMount() {
        this.fetchPlanets();

        this.setState({
            lifeCycle: 'componentDidMount',
        });
    }

    componentDidUpdate(prevProps, prevState) {
        if (
            prevState.planets &&
            prevState.planets.length !== this.state.planets.length
        ) {
            this.setState({
                lifeCycle: 'componentDidUpdate',
            });
        }
    }

    fetchPlanets() {
        SwapiService.getPlanets().then(response => {
            this.setState({
                isLoading: false,
                planets: response.data.results,
            });
            this.alderaan = response.data.results.find(
                p => p.name == 'Alderaan',
            );
        });
    }

    renderPlanetListItmes(planets) {
        return planets.map(p => (
            <li key={p.name} data-testid={this.kebabCase(p.name)}>
                {p.name}
            </li>
        ));
    }

    hasAlderaan(planets) {
        if (planets && planets.length > 0) {
            return planets.find(p => p.name == 'Alderaan') ? true : false;
        } else {
            return false;
        }
    }

    toggleAlderaan = planets => {
        if (this.hasAlderaan(planets)) {
            planets = planets.filter(p => p.name !== this.alderaan.name);
        } else {
            planets.push(this.alderaan);
        }

        this.setState({ planets });
    };

    kebabCase(string) {
        let result = string;

        // Convert camelCase capitals to kebab-case.
        result = result.replace(/([a-z][A-Z])/g, function(match) {
            return match.substr(0, 1) + '-' + match.substr(1, 1).toLowerCase();
        });

        // Convert non-camelCase capitals to lowercase.
        result = result.toLowerCase();

        // Convert non-alphanumeric characters to hyphens
        result = result.replace(/[^-a-z0-9]+/g, '-');

        // Remove hyphens from both ends
        result = result.replace(/^-+/, '').replace(/-$/, '');

        return result;
    }

    render() {
        const { planets, isLoading, lifeCycle } = this.state;
        let hasAlderaan = this.hasAlderaan(planets);

        return (
            <div className="planet-container">
                <div className="planet-list">
                    {!isLoading && planets ? (
                        <ul>{this.renderPlanetListItmes(planets)}</ul>
                    ) : (
                        <div>loading planets</div>
                    )}
                </div>

                <div className="life-cycles">
                    {!isLoading && planets && (
                        <button
                            type="button"
                            className={`btn btn-${
                                hasAlderaan ? 'danger' : 'success'
                            }`}
                            onClick={e => this.toggleAlderaan(planets)}
                        >
                            {hasAlderaan ? 'remove' : 'add'} Alderaan
                        </button>
                    )}
                    <p className="life-cycle">{lifeCycle}</p>
                </div>
            </div>
        );
    }
}
