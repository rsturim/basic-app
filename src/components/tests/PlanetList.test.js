import React from 'react';
import toJson from 'enzyme-to-json';

import PlanetList from '../PlanetList';

import { planets as mockPlanets } from './mockData/data';

/* ==  "in-file" example of using the enzyme-adapter for react */
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('A test', () => {
    test('verify the planet list is complete', () => {
        const wrapper = mount(<PlanetList />);

        expect(wrapper.find('.planet-list > div').text()).toEqual(
            'loading planets',
        );

        wrapper.setState({
            isLoading: false,
            planets: mockPlanets,
        });

        /* == verify Endor */
        expect(wrapper.find('li').length).toEqual(10);
        expect(wrapper.find('[data-testid="endor"]').text()).toEqual('Endor');

        let endorHtml = wrapper.find('[data-testid="endor"]').html();
        let expectedHtml = '<li data-testid="endor">Endor</li>';

        expect(endorHtml).toEqual(expectedHtml);
    });

    test('verify setting state will change the list of planets', () => {
        const wrapper = mount(<PlanetList />);

        /* == verify REMOVING Endor  BY Setting state*/
        let lessPlanets = mockPlanets.filter(planet => {
            return planet.name !== 'Endor';
        });

        wrapper.setState({
            isLoading: false,
            planets: lessPlanets,
        });

        expect(wrapper.find('li').length).toEqual(9);
        expect(wrapper.find('[data-testid="endor"]')).toEqual({});
        expect(wrapper.find('[data-testid="endor"]').values).toBe(undefined);
    });

    /* == componentDidMount */
    test('should run the componentDidMount lifecycle method', () => {
        jest.spyOn(PlanetList.prototype, 'componentDidMount');
        const wrapper = shallow(<PlanetList />);

        expect(PlanetList.prototype.componentDidMount.mock.calls.length).toBe(
            1,
        );

        expect(wrapper.find('.life-cycle').text()).toBe('componentDidMount');
    });

    /* == componentDidUpdate */
    test('should run the componentDidUpdate lifecycle method', () => {
        jest.spyOn(PlanetList.prototype, 'componentDidUpdate');
        const wrapper = shallow(<PlanetList />);

        wrapper.setState({
            isLoading: false,
            planets: mockPlanets,
        });

        expect(PlanetList.prototype.componentDidUpdate.mock.calls.length).toBe(
            3,
        );

        expect(wrapper.find('.life-cycle').text()).toBe('componentDidUpdate');
    });

    test('can kebabCase a string correctly', () => {
        const wrapper = shallow(<PlanetList />);

        const helloWorldTest = wrapper.instance().kebabCase('Hello World');
        expect(helloWorldTest).toBe('hello-world');

        const falseResult = wrapper.instance().kebabCase('');
        expect(falseResult).toBe('');
    });

    test('should match the snapshot', () => {
        const wrapper = mount(<PlanetList />);

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
