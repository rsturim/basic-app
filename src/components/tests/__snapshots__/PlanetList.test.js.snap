// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A test should match the snapshot 1`] = `
<PlanetList>
  <div
    className="planet-container"
  >
    <div
      className="planet-list"
    >
      <div>
        loading planets
      </div>
    </div>
    <div
      className="life-cycles"
    >
      <p
        className="life-cycle"
      >
        componentDidMount
      </p>
    </div>
  </div>
</PlanetList>
`;
